### ISSUE TITLE
incokeCommandCallback exploit

### ISSUE THREAD

**Issue by oligriffiths** [timble/kodekit#123]
`invokeCommandCallback` in `ControllerAbstract` allows for executing ANY private or protected method on the object: https://github.com/timble/kodekit/blob/master/code/controller/abstract.php#L156 without any check that the relevant callback is genuine. This should probably have some way of validating that the callback is valid.

**Comment by johanjanssens** [comment]
@oligriffiths Yes you could add additional checks, you would then need to take the mixed methods into account too. However if the method doesn't exists a BadMethodCallException will be thrown. See: https://github.com/timble/kodekit/blob/master/code/object/object.php#L444

Unless you want to further specialise this exception with a BadCallbackCallException this should be sufficient?

**Comment by oligriffiths** [comment]
@johanjanssens Less concerned about methods that don't exist, and more methods that do exist but are not registered callbacks. This currently allows a "backdoor" into any object to call protected/private methods. Will investigate a fix.

**Comment by johanjanssens** [comment]
I see what you mean. Because invokeCommandCallback can be called public-ally anyone could call it?
The same problem exists here: https://github.com/timble/kodekit/blob/5125e8f44b6b45bb32c0da9e720d1ba25ae01092/code/command/mixin/mixin.php#L151

To solve that we could add a hasCommandCallback in the [CommandCallbackInterface](https://github.com/timble/kodekit/blob/5125e8f44b6b45bb32c0da9e720d1ba25ae01092/code/command/callback/interface.php) which can check if a specific callback exists and only then call it. We can then implement checks in various places. Thoughts?

**Comment by oligriffiths** [comment]
Exactly.

Yeah, seems like a reasonable solution to me.

**Comment by johanjanssens** [comment]
Wanna work on this, or leave it to me?

**Comment by oligriffiths** [comment]
Happy to have a stab.

**Comment by johanjanssens** [comment]
All yours then, make it sexy! :)
